### ISSUE TITLE
sfdx suffers from EPROTO errors in at least one secured environment

### ISSUE THREAD

**Issue by mikecurwen** [forcedotcom/cli#1857]
### Summary
 
In certain network environments, newer versions of sfdx cannot make outbound connections
 
### Steps To Reproduce:
 
1. Install some version of sfdx prior to node upgrading to OpenSSL 3; in our case, that was 7.126 (using node 16.13)
2. sfdx update  (and when I tried that this past week or so, I got sfdx update to 7.181 at node 18.12.1)
 
### Expected result
 
sfdx ought to be updated to a new working version
 
### Actual result
 
What happens in our secured environment is sfdx (when attempting to connect somewhere external) gives:
```
sfdx-cli: Updating CLI... !
    Error: write EPROTO 4C200000:error:0A000152:SSL routines:final_renegotiate:unsafe legacy renegotiation
    disabled:c:\ws\deps\openssl\openssl\ssl\statem\extensions.c:922:
 
    Code: EPROTO
 
```
 
### System Information
 
```
{
  "cliVersion": "sfdx-cli/7.181.1",
  "architecture": "win32-x64",
  "nodeVersion": "node-v18.12.1",
  "pluginVersions": [
    "@oclif/plugin-autocomplete 1.3.6 (core)",
    "@oclif/plugin-commands 2.2.1 (core)",
    "@oclif/plugin-help 5.1.19 (core)",
    "@oclif/plugin-not-found 2.3.9 (core)",
    "@oclif/plugin-plugins 2.1.7 (core)",
    "@oclif/plugin-search 0.0.4 (core)",
    "@oclif/plugin-update 3.0.7 (core)",
    "@oclif/plugin-version 1.1.3 (core)",
    "@oclif/plugin-warn-if-update-available 2.0.14 (core)",
    "@oclif/plugin-which 2.2.2 (core)",
    "alias 2.1.11 (core)",
    "apex 1.3.0 (core)",
    "auth 2.3.1 (core)",
    "community 2.0.18 (core)",
    "config 1.4.22 (core)",
    "custom-metadata 2.0.5 (core)",
    "data 2.1.15 (core)",
    "generator 2.0.12 (core)",
    "info 2.2.11 (core)",
    "limits 2.1.3 (core)",
    "org 2.2.18 (core)",
    "packaging 1.10.10 (core)",
    "schema 2.1.18 (core)",
    "signups 1.2.12 (core)",
    "source 2.3.7 (core)",
    "telemetry 2.0.5 (core)",
    "templates 55.1.0 (core)",
    "trust 2.2.1 (core)",
    "user 2.1.22 (core)",
    "@salesforce/sfdx-plugin-lwc-test 1.0.1 (core)",
    "salesforce-alm 54.8.4 (core)"
  ],
  "osVersion": "Windows_NT 10.0.19044",
  "shell": "cmd.exe",
  "rootPath": "C:\\Users\\mcurwen\\AppData\\Local\\sfdx\\client\\7.181.1-05b8719"
}
```
### Additional information
 
my own 'analysis' so far is that this is because sfdx began using a version of node (18) that began using a version of OpenSSL (3.0) that is no longer cool with "unsafe legacy renegotiation".  Trying this update on my home system works fine.  Trying it on my work system fails 100% of the time.  And by trying "this" i mean anything (once you've updated past the magic version number).  sfdx update, authenticating to an org, etc. Anything going through our Palo Alto, which does that decryption/re-encryption MITM so it can sniff traffic.  I think OpenSSL asks "do you support secure reneg?" and our Palo is responding "no" because it might not support either unsafe legacy, or new secured reneg. https://live.paloaltonetworks.com/t5/general-topics/ssl-routines-unsafe-legacy-renegotiation-disabled/td-p/520652

I will likely have to go through our Security Operations process to beg permission to turn that off for any URL that sfdx might interact with.  Which might be a bit of a lift.  So, in the meantime, with my background as a programmer (although not a SF or node developer) I wondered if there were some buttons I could push...
 
I didn't find any environment variables of the SFDX_* variety that got me past this OpenSSL3 hump.  I also didn't really find node params to directly affect this, although I did find --openssl-config=<path> as part of this one/two series of pages:
https://stackoverflow.com/questions/74324019/allow-legacy-renegotiation-for-nodejs
and
https://www.mail-archive.com/openssl-users@openssl.org/msg90598.html
 
Just like the OP of the stackoverflow post - I couldn't get the openSSL config options to work for me.  So the OpenSSL vars I discovered (SSL_OP_NO_RENEGOTIATION, SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION, SSL_OP_LEGACY_SERVER_CONNECT) are tantalizing close, but also so far away.

Is there any hints you can send my way?  or perhaps (hope beyond hope) - would you consider adding a SFDX_OVERCOME_LEGACY_RENEG_ISSUE_IN_SECURE_ENVIRONMENTS variable?

**Comment by cristiand391** [comment]
Hmm, have you tried passing `--openssl-legacy-provider` to the node process?
https://github.com/nodejs/node/blob/main/doc/changelogs/CHANGELOG_V17.md#17.0.0

You can pass these flags to the node process through the `NODE_OPTIONS` env var:

`NODE_OPTIONS='--openssl-legacy-provider' sfdx <command>`

Our installers bundle node so you don't have to worry about it, but you could also use node v16 and install sfdx from npm, that should work for you (at least until node v16 is EOL'd), see:
https://developer.salesforce.com/docs/atlas.en-us.sfdx_setup.meta/sfdx_setup/sfdx_setup_install_cli.htm#sfdx_setup_install_cli_npm

**Comment by mikecurwen** [comment]
It doesn't work. 

I had found that parameter in my investigation and not tried it because the error was not the same and reading up here: https://www.openssl.org/docs/man3.0/man7/OSSL_PROVIDER-legacy.html  it appeared to mainly talk about legacy crypto algorithms. Whereas my issue is to do with new openSSL behaviour to enforce RFC 5746 and the specific switch to turn off that behaviour is that SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION (which node doesn't have a --param for).  

But let's try it anyways, maybe it will work.    And alas, it doesn't. 

as for "you could use node v16 and install from npm"...  in our security environment, I discovered that installing node on windows is an agony of painful "and now I need this URL unblocked" requests.  Nothing is enabled or opened by default.  *.exe's all need to be whitelisted, and installed via a "software center" app, not by double-clicking an *.exe you downloaded yourself from the internet (not that you could, because the firewall would block them). Software that updates itself over the internet is anathema.

**Comment by mshanemc** [comment]
looks like PaloAlto has about 9 months to get this all worked out: 

https://nodejs.org/en/blog/announcements/nodejs16-eol/

**Comment by mshanemc** [comment]
I'm going to mark this as `wontfix` because there's nothing we can do about node's internals.  Just in case they're not already, you probably want to make sure the vendor (PaloAlto, or maybe others) are aware of the issue and working on a fix.

In the short term, it sounds like you'll either need some exceptions from your Security team, or finding a way to use node16 (either getting it installed or using an older CLI version from before node18 became LTS).

**Comment by mikecurwen** [comment]
I was hoping that an sfdx dev who could contemplate the code found on this page https://stackoverflow.com/questions/74324019/allow-legacy-renegotiation-for-nodejs and see if there's something similar that would work for sfdx.  Entirely unsurprisingly, that code didn't work for me when I hacked it in somewhere (I did do a console.log to make sure I had a js file that was being run). 

Of the two paths discussed in that post (configuration vs code) I did  try the configuration first, but like the OP on the stackoverflow, I could not get it to work.    If there's no interest in looking at a code fix - does anyone have suggestions on a config fix?  I suppose I should really be asking in a node or openssl forum, but I'm taking all ideas.

**Comment by cristiand391** [comment]
@mikecurwen I haven't found any other config solution for this.

Even if we could add some config in our HTTP client and expose it through an env var, it's not guaranteed that all plugins will use it (some 3rd party plugin may use axios/got/node-fetch instead of ours) so it wouldn't work (and also it wouldn't be a fix but a patch as the issue is on the server side).
